In [14]:
import pandas as pd
#!pip install sidetable
import sidetable

### First block: 
- In which district are there more accidents? 
- How does it relate to population? 
- What are the most common causes by district?
- What is the average number of victims and or vehicules implicated by district?



In [83]:
#import and check de dataframe of victims.
path = r'C:\Users\Usuario\Desktop\gits\Project-Week-2-Barcelona\datasets'
dataset_acc = pd.read_csv(path+'/'+'2019_accidents_gu_bcn.csv')
dataset_acc

Numero_expedient  Codi_districte        Nom_districte  Codi_barri  \
0      2019S000001                   5  Sarrià-Sant Gervasi          26   
1      2019S000002                   9          Sant Andreu          62   
2      2019S000003                  10           Sant Martí          66   
3      2019S000004                   6               Gràcia          32   
4      2019S000005                   2             Eixample           7   
...                ...             ...                  ...         ...   
10022  2019S010046                   4            Les Corts          21   
10023  2019S010047                  10           Sant Martí          64   
10024  2019S010048                   2             Eixample           5   
10025  2019S010049                   4            Les Corts          19   
10026  2019S010050                   2             Eixample           6   

                                Nom_barri  Codi_carrer  \
0                  Sant Gervasi - Galvany       144601   
1                el Congrés i els Indians       102907   
2       el Parc i la Llacuna del Poblenou       242906   
3      el Camp d'en Grassot i Gràcia Nova       228803   
4                  la Dreta de l'Eixample        89004   
...                                   ...          ...   
10022                           Pedralbes       101700   
10023          el Camp de l'Arpa del Clot       297001   
10024                       el Fort Pienc        28305   
10025                           les Corts       167800   
10026                  la Sagrada Família       178308   

                                              Nom_carrer Num_postal_caption  \
0      Diagonal / Augusta                            ...          0482 0482   
1      Felip II / Congrés Eucarístic                 ...          9999 9999   
2      Pallars                                       ...          0111 0113   
3      Taxdirt / Nogués                              ...          0054 0054   
4      Consell de Cent / Girona                      ...          0395 0397   
...                                                  ...                ...   
10022  Doctor Ferran / Manila                        ...          0025 0025   
10023  Sant Antoni Maria Claret                      ...          0346 0348   
10024  Ausiàs Marc / Nàpols                          ...          0080 0080   
10025  Joaquim Molins                                ...          0009 0009   
10026  Lepant                                        ...          0255 0255   

      Descripcio_dia_setmana Dia_setmana  ...  \
0                    Dimarts          Dm  ...   
1                    Dimarts          Dm  ...   
2                    Dimarts          Dm  ...   
3                    Dimarts          Dm  ...   
4                    Dimarts          Dm  ...   
...                      ...         ...  ...   
10022                Dimarts          Dm  ...   
10023                Dimarts          Dm  ...   
10024                Dimarts          Dm  ...   
10025                Dimarts          Dm  ...   
10026               Dimecres          Dc  ...   

              Descripcio_causa_vianant  Numero_morts  Numero_lesionats_lleus  \
0             No és causa del  vianant             0                       0   
1             No és causa del  vianant             0                       1   
2      Creuar per fora pas de vianants             0                       1   
3             No és causa del  vianant             0                       1   
4             No és causa del  vianant             0                       2   
...                                ...           ...                     ...   
10022         No és causa del  vianant             0                       1   
10023         No és causa del  vianant             0                       1   
10024         No és causa del  vianant             0                       1   
10025         No és causa del  vianant             0                  

Let's see what is the total of accidents in Barcelona in 2019

In [23]:
num_accidents = dataset_acc.shape[0]
num_accidents

10027

We need to count the number of ocurrences of each district, this will be easy because we have the variable both encoded and decoded (string).

We use sidetable for this, which is a sublibrary of pandas. However since we are going to use it later for merging and we need it to have the rows in the same order we are going to rearrange the ones from the following dataset, so we'll do it by the order in the list.

In [24]:
acc_district = dataset_acc.stb.freq(['Nom_districte'])
acc_district

Nom_districte  Count   Percent  Cumulative Count  Cumulative Percent
0              Eixample   3010  0.300189              3010            0.300189
1            Sant Martí   1243  0.123965              4253            0.424155
2        Sants-Montjuïc   1172  0.116884              5425            0.541039
3   Sarrià-Sant Gervasi   1053  0.105016              6478            0.646056
4             Les Corts    668  0.066620              7146            0.712676
5        Horta-Guinardó    664  0.066221              7810            0.778897
6           Sant Andreu    643  0.064127              8453            0.843024
7          Ciutat Vella    564  0.056248              9017            0.899272
8            Nou Barris    528  0.052658              9545            0.951930
9                Gràcia    463  0.046175             10008            0.998105
10           Desconegut     19  0.001895             10027            1.000000

Now Im gonna import the table that I found in this source https://www.bcn.cat/estadistica/castella/dades/inf/lecpadro/lec19/t12.htm which is population by district to gain more insights

In [25]:
dataset_pop_dist = pd.read_csv(path+'/'+'poblacio_districte.csv')

This one we do need to clean, but will be worth it because we will use it to answer the following questions too.

In [26]:
dataset_pop_dist1 = dataset_pop_dist.copy()
data_dist2 = dataset_pop_dist1.drop(['Unnamed: 1'], axis=1)

We also drop the % column

In [27]:
data_dist3 = data_dist2.drop(['%'], axis=1)

In [28]:
def cleaner_str(x):
    for index,i in enumerate(x):
        if x != 'BARCELONA':
            new_i = x[3:]
        
        else:
            new_i = x
    return new_i

#Applying the variable to each row in the column Distritos
data_dist3['Distritos'] = data_dist3['Distritos'].apply(cleaner_str)
data_dist3

Distritos      TOTAL  Hombres   %.1  Mujeres   %.2
0             BARCELONA  1.650.358  782.555  47,4  867.803  52,6
1          Ciutat Vella    105.820   56.107    53   49.713    47
2              Eixample    269.095  125.612  46,7  143.483  53,3
3        Sants-Montjuïc    185.450   88.853  47,9   96.597  52,1
4             Les Corts     82.591   38.367  46,5   44.224  53,5
5   Sarrià-Sant Gervasi    150.446   69.045  45,9   81.401  54,1
6                Gràcia    122.853   56.404  45,9   66.449  54,1
7        Horta-Guinardó    172.473   80.937  46,9   91.536  53,1
8            Nou Barris    171.290   80.371  46,9   90.919  53,1
9           Sant Andreu    150.264   71.335  47,5   78.929  52,5
10           Sant Martí    240.076  115.524  48,1  124.552  51,9

In [29]:
data_dist3.to_csv(path+'/'+'poblacio_edited.csv')

Since we could not manage to rearrange the rows we opted for doing it on excel ( though it makes me mad because I wanna learn to use this) ASK GUILLEM

In [30]:
dataset_pop_dist2 = pd.read_csv(path+'/'+'poblacio_edited2 - poblacio_edited.csv')


In [31]:
#We drom the Unnamed column
dataset_pop_dist2 = dataset_pop_dist2.drop('Unnamed: 0', axis= 1)
dataset_pop_dist2

Distritos      TOTAL  Hombres   %.1  Mujeres   %.2
0             BARCELONA  1.650.358  782.555  47,4  867.803  52,6
1              Eixample    269.095  125.612  46,7  143.483  53,3
2            Sant Martí    240.076  115.524  48,1  124.552  51,9
3        Sants-Montjuïc    185.450   88.853  47,9   96.597  52,1
4   Sarrià-Sant Gervasi    150.446   69.045  45,9   81.401  54,1
5             Les Corts     82.591   38.367  46,5   44.224  53,5
6        Horta-Guinardó    172.473   80.937  46,9   91.536  53,1
7           Sant Andreu    150.264   71.335  47,5   78.929  52,5
8          Ciutat Vella    105.820   56.107    53   49.713    47
9            Nou Barris    171.290   80.371  46,9   90.919  53,1
10               Gràcia    122.853   56.404  45,9   66.449  54,1

In [32]:
#We rename the column to do the merge on it and also replace the nme Sant Marti because I tried the merge once and it didnt catch it
#witch means it has something different.Its the order!
acc_district.rename(columns={'Nom_districte':'Distritos'}, inplace=True)

In [33]:
merged_district =acc_district.merge(right=dataset_pop_dist2, how='inner', on='Distritos')
merged_district

Distritos  Count   Percent  Cumulative Count  Cumulative Percent  \
0             Eixample   3010  0.300189              3010            0.300189   
1           Sant Martí   1243  0.123965              4253            0.424155   
2       Sants-Montjuïc   1172  0.116884              5425            0.541039   
3  Sarrià-Sant Gervasi   1053  0.105016              6478            0.646056   
4            Les Corts    668  0.066620              7146            0.712676   
5       Horta-Guinardó    664  0.066221              7810            0.778897   
6          Sant Andreu    643  0.064127              8453            0.843024   
7         Ciutat Vella    564  0.056248              9017            0.899272   
8           Nou Barris    528  0.052658              9545            0.951930   
9               Gràcia    463  0.046175             10008            0.998105   

     TOTAL  Hombres   %.1  Mujeres   %.2  
0  269.095  125.612  46,7  143.483  53,3  
1  240.076  115.524  48,1  124.552  51,9  
2  185.450   88.853  47,9   96.597  52,1  
3  150.446   69.045  45,9   81.401  54,1  
4   82.591   38.367  46,5   44.224  53,5  
5  172.473   80.937  46,9   91.536  53,1  
6  150.264   71.335  47,5   78.929  52,5  
7  105.820   56.107    53   49.713    47  
8  171.290   80.371  46,9   90.919  53,1  
9  122.853   56.404  45,9   66.449  54,1

In [34]:
#We change some column names to make it more readable.
merged_district = merged_district.rename(columns={'Percent':'Percent_acci','Cumulative Count':'Cumulative_Count','Cumulative Percent':'Cumulative_Percent','TOTAL':'Total_Pop'})

In [35]:
#We have to replace the ',' in %.1 and %.2 by '.' to convert into float
merged_district['%.1'] = merged_district['%.1'].apply(lambda x: x.replace(',', '.')).astype('float')
merged_district['%.2'] = merged_district['%.2'].apply(lambda x: x.replace(',', '.')).astype('float')
merged_district

Distritos  Count  Percent_acci  Cumulative_Count  \
0             Eixample   3010      0.300189              3010   
1           Sant Martí   1243      0.123965              4253   
2       Sants-Montjuïc   1172      0.116884              5425   
3  Sarrià-Sant Gervasi   1053      0.105016              6478   
4            Les Corts    668      0.066620              7146   
5       Horta-Guinardó    664      0.066221              7810   
6          Sant Andreu    643      0.064127              8453   
7         Ciutat Vella    564      0.056248              9017   
8           Nou Barris    528      0.052658              9545   
9               Gràcia    463      0.046175             10008   

   Cumulative_Percent Total_Pop  Hombres   %.1  Mujeres   %.2  
0            0.300189   269.095  125.612  46.7  143.483  53.3  
1            0.424155   240.076  115.524  48.1  124.552  51.9  
2            0.541039   185.450   88.853  47.9   96.597  52.1  
3            0.646056   150.446   69.045  45.9   81.401  54.1  
4            0.712676    82.591   38.367  46.5   44.224  53.5  
5            0.778897   172.473   80.937  46.9   91.536  53.1  
6            0.843024   150.264   71.335  47.5   78.929  52.5  
7            0.899272   105.820   56.107  53.0   49.713  47.0  
8            0.951930   171.290   80.371  46.9   90.919  53.1  
9            0.998105   122.853   56.404  45.9   66.449  54.1

In [36]:
#We make sure all numeric columns are set to int or float.
cols =['Percent_acci','Cumulative_Percent','Total_Pop','Hombres','Mujeres']
merged_district[cols] = merged_district[cols].astype(float)

In [37]:
#We create a column Pop_Percent to know which percentatge of the population represents each district 
merged_district['Pop_Percent'] = ((merged_district['Total_Pop'] / merged_district['Total_Pop'].sum())*100).round(2)
merged_district

Distritos  Count  Percent_acci  Cumulative_Count  \
0             Eixample   3010      0.300189              3010   
1           Sant Martí   1243      0.123965              4253   
2       Sants-Montjuïc   1172      0.116884              5425   
3  Sarrià-Sant Gervasi   1053      0.105016              6478   
4            Les Corts    668      0.066620              7146   
5       Horta-Guinardó    664      0.066221              7810   
6          Sant Andreu    643      0.064127              8453   
7         Ciutat Vella    564      0.056248              9017   
8           Nou Barris    528      0.052658              9545   
9               Gràcia    463      0.046175             10008   

   Cumulative_Percent  Total_Pop  Hombres   %.1  Mujeres   %.2  Pop_Percent  
0            0.300189    269.095  125.612  46.7  143.483  53.3        16.31  
1            0.424155    240.076  115.524  48.1  124.552  51.9        14.55  
2            0.541039    185.450   88.853  47.9   96.597  52.1        11.24  
3            0.646056    150.446   69.045  45.9   81.401  54.1         9.12  
4            0.712676     82.591   38.367  46.5   44.224  53.5         5.00  
5            0.778897    172.473   80.937  46.9   91.536  53.1        10.45  
6            0.843024    150.264   71.335  47.5   78.929  52.5         9.10  
7            0.899272    105.820   56.107  53.0   49.713  47.0         6.41  
8            0.951930    171.290   80.371  46.9   90.919  53.1        10.38  
9            0.998105    122.853   56.404  45.9   66.449  54.1         7.44

In [38]:
#We compute the % of population this 3 districts accumulate
merged_district.iloc[:3, [10]].sum()

Pop_Percent    42.1
dtype: float64

In [39]:
#We compute the cumulated count of accidents for the three with less acc
(merged_district.iloc[7:,[1]].sum()/merged_district.iloc[:,[1]].sum())*100

Count    15.53757
dtype: float64

In [40]:
#We compute the cumulated perc of pop  in the three with less accidents
(merged_district.iloc[7:,[5]].sum()/merged_district.iloc[:,[5]].sum())*100

Total_Pop    24.234924
dtype: float64

In [41]:
#% combined pop of the outliers
(merged_district.iloc[5:7,[5]].sum()/merged_district.iloc[:,[5]].sum())*100

Total_Pop    19.555575
dtype: float64

In [42]:
#% accidents of the outliers
(merged_district.iloc[5:7,[1]].sum()/merged_district.iloc[:,[1]].sum())*100

Count    13.059552
dtype: float64

In [43]:
merged_district.corr()

Count  Percent_acci  Cumulative_Count  \
Count               1.000000      1.000000         -0.853110   
Percent_acci        1.000000      1.000000         -0.853110   
Cumulative_Count   -0.853110     -0.853110          1.000000   
Cumulative_Percent -0.853110     -0.853110          1.000000   
Total_Pop           0.782220      0.782220         -0.755987   
Hombres             0.770106      0.770106         -0.754778   
%.1                -0.156293     -0.156293          0.135563   
Mujeres             0.787682      0.787682         -0.752103   
%.2                 0.156293      0.156293         -0.135563   
Pop_Percent         0.782301      0.782301         -0.756145   

                    Cumulative_Percent  Total_Pop   Hombres       %.1  \
Count                        -0.853110   0.782220  0.770106 -0.156293   
Percent_acci                 -0.853110   0.782220  0.770106 -0.156293   
Cumulative_Count              1.000000  -0.755987 -0.754778  0.135563   
Cumulative_Percent            1.000000  -0.755987 -0.754778  0.135563   
Total_Pop                    -0.755987   1.000000  0.996002 -0.179572   
Hombres                      -0.754778   0.996002  1.000000 -0.093431   
%.1                           0.135563  -0.179572 -0.093431  1.000000   
Mujeres                      -0.752103   0.996957  0.986006 -0.253569   
%.2                          -0.135563   0.179572  0.093431 -1.000000   
Pop_Percent                  -0.756145   1.000000  0.996002 -0.179539   

                     Mujeres       %.2  Pop_Percent  
Count               0.787682  0.156293     0.782301  
Percent_acci        0.787682  0.156293     0.782301  
Cumulative_Count   -0.752103 -0.135563    -0.756145  
Cumulative_Percent -0.752103 -0.135563    -0.756145  
Total_Pop           0.996957  0.179572     1.000000  
Hombres             0.986006  0.093431     0.996002  
%.1                -0.253569 -1.000000    -0.179539  
Mujeres             1.000000  0.253569     0.996956  
%.2                 0.253569  1.000000     0.179539  
Pop_Percent         0.996956  0.179539     1.000000

In [44]:
merged_district.describe()

Count  Percent_acci  Cumulative_Count  Cumulative_Percent  \
count    10.00000     10.000000         10.000000           10.000000   
mean   1000.80000      0.099811       7114.500000            0.709534   
std     758.89932      0.075686       2320.682814            0.231443   
min     463.00000      0.046175       3010.000000            0.300189   
25%     583.75000      0.058218       5688.250000            0.567293   
50%     666.00000      0.066421       7478.000000            0.745786   
75%    1142.25000      0.113917       8876.000000            0.885210   
max    3010.00000      0.300189      10008.000000            0.998105   

        Total_Pop     Hombres       %.1     Mujeres       %.2  Pop_Percent  
count   10.000000   10.000000  10.00000   10.000000  10.00000    10.000000  
mean   165.035800   78.255500  47.53000   86.780300  52.47000    10.000000  
std     57.236962   26.765266   2.06131   30.672057   2.06131     3.470978  
min     82.591000   38.367000  45.90000   44.224000  47.00000     5.000000  
25%    129.705750   59.564250  46.55000   69.569000  52.20000     7.855000  
50%    160.868000   75.853000  46.90000   86.160000  53.10000     9.750000  
75%    182.205750   86.874000  47.80000   95.331750  53.45000    11.042500  
max    269.095000  125.612000  53.00000  143.483000  54.10000    16.310000

### Insights: 

To the questions :
- In which district are there more accidents? 

The 54% of accidents happen in just 3 districts (Eixample, St Marti, Montjuic) that sum the 42.1% of the population.The 3 district with the less accidents accumulate 15.53% of them and its population percentage is  24.23%. Remarkably Horta and St Andreu are low on accidents for their level of population , which is 19.55% combined and their percentage of accidents is 13.06%.

- How does it relate to population? 

The level of population(of course both absolute and perccentual) is very positively related to the number of accidents, as expected.

We have answered the first 2 questions and will attack the next 2 with the main dataframe, from which we constructed teh cumulative one, the original *dataset_acc

In [45]:
dataset_acc.columns

Index(['Numero_expedient', 'Codi_districte', 'Nom_districte', 'Codi_barri',
       'Nom_barri', 'Codi_carrer', 'Nom_carrer', 'Num_postal_caption',
       'Descripcio_dia_setmana', 'Dia_setmana', 'Descripcio_tipus_dia',
       'NK_Any', 'Mes_any', 'Nom_mes', 'Dia_mes', 'Hora_dia',
       'Descripcio_torn', 'Descripcio_causa_vianant', 'Numero_morts',
       'Numero_lesionats_lleus', 'Numero_lesionats_greus', 'Numero_victimes',
       'Numero_vehicles_implicats', 'Coordenada_UTM_X', 'Coordenada_UTM_Y',
       'Longitud', 'Latitud'],
      dtype='object')

In [46]:
#Drop the columns we are not going to use in our analysis
dataset_acc= dataset_acc.drop(['Numero_expedient', 'Codi_barri',
       'Nom_barri', 'Codi_carrer', 'Nom_carrer', 'Num_postal_caption',
       'Descripcio_dia_setmana', 'Dia_setmana', 'Descripcio_tipus_dia',
       'NK_Any', 'Mes_any', 'Nom_mes', 'Dia_mes', 'Hora_dia',
       'Descripcio_torn', 'Coordenada_UTM_X', 'Coordenada_UTM_Y',
       'Longitud', 'Latitud'], axis=1)

In [47]:
dataset_acc = dataset_acc.drop(['Codi_districte'],axis =1)


In [48]:
dataset_acc

Nom_districte         Descripcio_causa_vianant  Numero_morts  \
0      Sarrià-Sant Gervasi         No és causa del  vianant             0   
1              Sant Andreu         No és causa del  vianant             0   
2               Sant Martí  Creuar per fora pas de vianants             0   
3                   Gràcia         No és causa del  vianant             0   
4                 Eixample         No és causa del  vianant             0   
...                    ...                              ...           ...   
10022            Les Corts         No és causa del  vianant             0   
10023           Sant Martí         No és causa del  vianant             0   
10024             Eixample         No és causa del  vianant             0   
10025            Les Corts         No és causa del  vianant             0   
10026             Eixample         No és causa del  vianant             0   

       Numero_lesionats_lleus  Numero_lesionats_greus  Numero_victimes  \
0                           0                       1                1   
1                           1                       0                1   
2                           1                       0                1   
3                           1                       0                1   
4                           2                       0                2   
...                       ...                     ...              ...   
10022                       1                       0                1   
10023                       1                       0                1   
10024                       1                       0                1   
10025                       1                       0                1   
10026                       0                       1                1   

       Numero_vehicles_implicats  
0                              1  
1                              2  
2                              1  
3                              2  
4                              1  
...                          ...  
10022                          1  
10023                          2  
10024                          2  
10025                          2  
10026                          2  

[10027 rows x 7 columns]

In [54]:
#We groupby district and the cause as descripted by the pedestrian and count the number of accidents 
cause_by_district = dataset_acc.groupby(['Nom_districte','Descripcio_causa_vianant'])['Descripcio_causa_vianant'].count()
#Turn it into a dataframe so we can sumirize things and gain insights
cause_by_district= pd.DataFrame(cause_by_district)
cause_by_district= cause_by_district.rename(columns={'Descripcio_causa_vianant':'Accidents'})
cause_by_district

Accidents
Nom_districte       Descripcio_causa_vianant                  
Ciutat Vella        Altres                                  19
                    Creuar per fora pas de vianants         20
                    Desobeir el senyal del semàfor           9
                    No és causa del  vianant               508
                    Transitar a peu per la calçada           8
Desconegut          Altres                                   2
                    Creuar per fora pas de vianants          1
                    No és causa del  vianant                16
Eixample            Altres                                  44
                    Creuar per fora pas de vianants         68
                    Desobeir altres senyals                  1
                    Desobeir el senyal del semàfor          92
                    No és causa del  vianant              2792
                    Transitar a peu per la calçada          13
Gràcia              Altres                                  10
                    Creuar per fora pas de vianants          7
                    Desobeir el senyal del semàfor           8
                    No és causa del  vianant               436
                    Transitar a peu per la calçada           2
Horta-Guinardó      Altres                                   7
                    Creuar per fora pas de vianants         19
                    Desobeir el senyal del semàfor          11
                    No és causa del  vianant               623
                    Transitar a peu per la calçada           4
Les Corts           Altres                                  10
                    Creuar per fora pas de vianants          7
                    Desobeir el senyal del semàfor          15
                    No és causa del  vianant               634
                    Transitar a peu per la calçada           2
Nou Barris          Altres                                  10
                    Creuar per fora pas de vianants         18
                    Desobeir el senyal del semàfor           8
                    No és causa del  vianant               490
                    Transitar a peu per la calçada           2
Sant Andreu         Altres                                   9
                    Creuar per fora pas de vianants         20
                    Desobeir el senyal del semàfor          17
                    No és causa del  vianant               597
Sant Martí          Altres                                  11
                    Creuar per fora pas de vianants         19
                    Desobeir altres senyals                  2
                    Desobeir el senyal del semàfor          21
                    No és causa del  vianant              1181
                    Transitar a peu per la calçada           9
Sants-Montjuïc      Altres                                  20
                    Creuar per fora pas de vianants         20
                    Desobeir altres senyals                  1
                    Desobeir el senyal del semàfor          31
                    No és causa del  vianant              1093
                    Transitar a peu per la calçada           7
Sarrià-Sant Gervasi Altres                                  14
                    Creuar per fora pas de vianants         32
                    Desobeir altres senyals                  1
                    Desobeir el senyal del semàfor          18
                    No és causa del  vianant               984
                    Transitar a peu per la calçada           4

### Insights

To the question:

- What are the most common causes by district?


We are not going to analyze all of them so we are going to focus on the top3 with most accidents and then the bottom least. 

Eixample: most common is the accident not caused by the pedestrian, which seems to be consistent across districts. The pedestrian skipping a stop light is second most common but way less than the first  one. The least common cause is the pedestrian disobeying other signals.

Sant Martí: Again most usually is not the pedestrian's fault, and almost tied in the secon position are disobeying a stop-light and not crossing through the crosswalk.Least common is the same as Eixample.

Sants-Montjuic:Same,then the stop-light. and again the least common is the same, disobeying other traffic signs.

Now lets quickly check the districts with less accidents.

Ciutat Vella: After being the drivers fault, not crossing through the corsswalk and 'Others' are the most common causes. and the sme virtually goes for Nou Barris and Gracia.

We conclude that the causes of an accident involving a pedestrian are more or less homogeneous across districts.

In [ ]:
dataset_acc.groupby(['Nom_districte',])

In [56]:
# Not many not worthy of analysis
dataset_acc.groupby(['Nom_districte','Descripcio_causa_vianant'])['Numero_morts'].sum()

Nom_districte        Descripcio_causa_vianant       
Ciutat Vella         Altres                             0
                     Creuar per fora pas de vianants    0
                     Desobeir el senyal del semàfor     0
                     No és causa del  vianant           2
                     Transitar a peu per la calçada     0
Desconegut           Altres                             0
                     Creuar per fora pas de vianants    0
                     No és causa del  vianant           0
Eixample             Altres                             0
                     Creuar per fora pas de vianants    1
                     Desobeir altres senyals            0
                     Desobeir el senyal del semàfor     0
                     No és causa del  vianant           2
                     Transitar a peu per la calçada     1
Gràcia               Altres                             0
                     Creuar per fora pas de vianants    0
                   

In [61]:
dataset_acc.groupby(['Nom_districte'], as_index = False)['Numero_vehicles_implicats'].max()

Nom_districte  Numero_vehicles_implicats
0          Ciutat Vella                          9
1            Desconegut                          4
2              Eixample                         18
3                Gràcia                          6
4        Horta-Guinardó                          6
5             Les Corts                          8
6            Nou Barris                         10
7           Sant Andreu                         10
8            Sant Martí                         18
9        Sants-Montjuïc                          7
10  Sarrià-Sant Gervasi                         11

In [62]:
dataset_acc.corr()

Numero_morts  Numero_lesionats_lleus  \
Numero_morts                   1.000000               -0.022981   
Numero_lesionats_lleus        -0.022981                1.000000   
Numero_lesionats_greus         0.022009               -0.114836   
Numero_victimes                0.040413                0.979830   
Numero_vehicles_implicats     -0.013591                0.172018   

                           Numero_lesionats_greus  Numero_victimes  \
Numero_morts                             0.022009         0.040413   
Numero_lesionats_lleus                  -0.114836         0.979830   
Numero_lesionats_greus                   1.000000         0.077072   
Numero_victimes                          0.077072         1.000000   
Numero_vehicles_implicats                0.000432         0.171832   

                           Numero_vehicles_implicats  
Numero_morts                               -0.013591  
Numero_lesionats_lleus                      0.172018  
Numero_lesionats_greus                      0.000432  
Numero_victimes                             0.171832  
Numero_vehicles_implicats                   1.000000

In [67]:
dataset_acc.groupby(['Nom_districte'], as_index = False)['Numero_vehicles_implicats','Numero_victimes'].mean()

C:\Users\Usuario\anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


Nom_districte  Numero_vehicles_implicats  Numero_victimes
0          Ciutat Vella                   1.746454         1.040780
1            Desconegut                   1.842105         1.210526
2              Eixample                   1.924585         1.197342
3                Gràcia                   1.831533         1.133909
4        Horta-Guinardó                   1.811747         1.180723
5             Les Corts                   1.931138         1.248503
6            Nou Barris                   1.863636         1.104167
7           Sant Andreu                   1.877138         1.230171
8            Sant Martí                   1.968624         1.206758
9        Sants-Montjuïc                   1.875427         1.175768
10  Sarrià-Sant Gervasi                   1.943970         1.191833

In [66]:
dataset_acc.groupby(['Nom_districte'], as_index = False)['Numero_vehicles_implicats','Numero_victimes'].max()

C:\Users\Usuario\anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


Nom_districte  Numero_vehicles_implicats  Numero_victimes
0          Ciutat Vella                          9               13
1            Desconegut                          4                3
2              Eixample                         18               26
3                Gràcia                          6                5
4        Horta-Guinardó                          6                7
5             Les Corts                          8                7
6            Nou Barris                         10                5
7           Sant Andreu                         10               14
8            Sant Martí                         18                9
9        Sants-Montjuïc                          7                7
10  Sarrià-Sant Gervasi                         11               12

### Insights

To the question:

- What is the average number of victims and or vehicules implicated by district?
This is shocking, although the mean of the veichles implicated in an accident is roughly constant across districts, the maximums are quite unstable and sexy, going as high as 18 vehicles for Eixample and Sant Marti (coincidentally 2 of the top3 trickier districts) but the shocking thing is that Nou Barris, one of the most chill districts had up to 10 vheicles in an accident. Let's relate it to the number of victims.

Turns out the number of victims is not too strongly correlated with the number of vheicles, at least on a general level.

We see on a more district by district level it is more closely correlated, at least the maximums are, since the higher it is the higher it also is for the number of victims. Though there are some cases like Nou Barris, or Sant Marti where the max number of vehicles doubles the maximum number of people, which means some veichles where either parked or where in the incident but had no victims. 

### Second Block:
 - What is the type of vehicle most implicated in accidents by district?
 - what brand/by type and district?
 - What color?

In [70]:
#We import the  vaihcles dataset
vei_type=pd.read_csv(path+'/'+'2019_accidents_vehicles_gu_bcn_.csv')
vei_type.columns

Index(['Numero_expedient', 'Codi_districte', 'Nom_districte', 'Codi_barri',
       'Nom_barri', 'Codi_carrer', 'Nom_carrer', 'Num_postal',
       'Descripcio_dia_setmana', 'Dia_setmana', 'Descripcio_tipus_dia',
       'NK_Any', 'Mes_any', 'Nom_mes', 'Dia_mes', 'Hora_dia',
       'Descripcio_torn', 'Descripcio_causa_vianant',
       'Descripcio_tipus_vehicle', 'Descripcio_model', 'Descripcio_marca',
       'Descripcio_color', 'Descripcio_carnet', 'Antiguitat_carnet',
       'Coordenada_UTM_X', 'Coordenada_UTM_Y', 'Longitud', 'Latitud'],
      dtype='object')

In [74]:
#We get rid of the columns we're not going to use.
vei_type = vei_type.drop(['Numero_expedient', 'Codi_districte', 'Codi_barri',
       'Nom_barri', 'Codi_carrer', 'Nom_carrer', 'Num_postal',
       'Descripcio_dia_setmana', 'Dia_setmana', 'Descripcio_tipus_dia',
       'NK_Any', 'Mes_any', 'Nom_mes', 'Dia_mes', 'Hora_dia',
       'Descripcio_torn', 'Descripcio_causa_vianant', 'Descripcio_carnet', 'Antiguitat_carnet',
       'Coordenada_UTM_X', 'Coordenada_UTM_Y', 'Longitud', 'Latitud'], axis = 1)

In [75]:
vei_type

Nom_districte Descripcio_tipus_vehicle Descripcio_model  \
0        Desconegut                Bicicleta       Desconegut   
1        Desconegut               Desconegut       Desconegut   
2        Desconegut              Motocicleta         WW125EX2   
3        Desconegut                  Turisme        307 SW PK   
4        Desconegut                  Turisme               C3   
...             ...                      ...              ...   
19032    Sant Martí                  Turisme          E 220 D   
19033    Sant Martí                  Turisme         SPORTAGE   
19034    Sant Martí                  Turisme            T-ROC   
19035    Sant Martí                  Turisme           MEGANE   
19036    Sant Martí              Motocicleta           SH 300   

      Descripcio_marca Descripcio_color  
0           Desconegut       Desconegut  
1           Desconegut       Desconegut  
2                HONDA            Blanc  
3              PEUGEOT             Gris  
4              CITROEN            Negre  
...                ...              ...  
19032    MERCEDES-BENZ       Negre/Groc  
19033              KIA          Vermell  
19034       VOLKSWAGEN            Blanc  
19035          RENAULT          Vermell  
19036            HONDA            Blanc  

[19037 rows x 5 columns]

In [81]:
import numpy as np
from scipy import stats

In [82]:
vei_type.groupby(['Nom_districte']).agg({"Descripcio_marca":stats.mode})

Descripcio_marca
Nom_districte                        
Ciutat Vella         ([HONDA], [101])
Desconegut           ([CITROEN], [3])
Eixample             ([HONDA], [911])
Gràcia               ([HONDA], [138])
Horta-Guinardó       ([HONDA], [172])
Les Corts            ([HONDA], [185])
Nou Barris            ([HONDA], [99])
Sant Andreu          ([HONDA], [132])
Sant Martí           ([HONDA], [279])
Sants-Montjuïc       ([HONDA], [247])
Sarrià-Sant Gervasi  ([HONDA], [366])